packages/tangle/scheduler.go

```diff
@@ -0,0 +1,298 @@
+package tangle
+
+import (
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/iotaledger/goshimmer/packages/clock"
+	"github.com/iotaledger/hive.go/async"
+	"github.com/iotaledger/hive.go/events"
+)
+
+var (
+	capacity   = 1000
+	numWorkers = runtime.NumCPU() * 4
+)
+
+// SchedulerParentPriorityMap maps parentIDs with their children messages.
+type SchedulerParentPriorityMap map[MessageID][]*MessageToSchedule
+
+// MessageToSchedule contains the relevant information to schedule a message.
+type MessageToSchedule struct {
+	ID          MessageID
+	parents     []MessageID
+	issuingTime time.Time
+	scheduled   bool
+}
+
+// Scheduler implements the scheduler.
+type Scheduler struct {
+	Events *SchedulerEvents
+
+	tangle *Tangle
+
+	inbox            chan *MessageToSchedule
+	timeQueue        *TimeMessageQueue
+	parentsMap       SchedulerParentPriorityMap
```

Comment:
should this be protected with a mutex?
we check on incoming message of parents have been booked, but the multithreaded outbox worker could modifiy the map in the meantime, no?

Reply:
No that map is managed entirely by the scheduler